File: core/src/main/java/org/zenika/core/PokemonUtils.java
package org.zenika.core;

import java.util.List;
import java.util.Random;

public class PokemonUtils {

    private static final int NORMAL = 2;
    private static final int NULL = 0;
    private static final int NOT_EFFECTIVE = 1;
    private static final int VERY_EFFECTIVE = 4;
    private static final Random random = new Random();

    public static final PokemonType pickType(Pokemon pokemon) {
        List<String> types = pokemon.getTypes();
        String pickedType = pick(types);
        return PokemonType.fromValue(pickedType);
    }

    public static final <T> T pick(List<T> givenList) {
        return givenList.get(random.nextInt(givenList.size()));
    }

    public static final int getAttackResult(Pokemon attacker, Pokemon defender) {
        PokemonType attackerType = pickType(attacker);
        PokemonType defenderType = pickType(defender);
        switch (attackerType) {
            case POISON:
                return getPoisonAttackResult(defenderType);
            case GRASS:
                return getGrassAttackResult(defenderType);
            case FIRE:
                return getFireAttackResult(defenderType);
            case FLYING:
                return getFlyingAttackResult(defenderType);
            case WATER:
                return getWaterAttackResult(defenderType);
            case BUG:
                return getBugAttackResult(defenderType);
            case NORMAL:
                return getNormalAttackResult(defenderType);
            case ELECTRIC:
                return getElectricAttackResult(defenderType);
            case GROUND:
                return getGroundAttackResult(defenderType);
            case FAIRY:
                return getFairyAttackResult(defenderType);
            case FIGHTING:
                return getFightingAttackResult(defenderType);
            case PSYCHIC:
                return getPsychicAttackResult(defenderType);
            case ROCK:
                return getRockAttackResult(defenderType);
            case STEEL:
                return getSteelAttackResult(defenderType);
            case ICE:
                return getIceAttackResult(defenderType);
            case GHOST:
                return getGhostAttackResult(defenderType);
            case DRAGON:
                return getDragonAttackResult(defenderType);
            case DARK:
                return getDarkAttackResult(defenderType);
        }
        return NORMAL;
    }

    public static final int getNormalAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case ROCK:
            case STEEL:
                return NOT_EFFECTIVE;
            case GHOST:
                return NULL;
            default: return NORMAL;
        }
    }

    public static final int getFireAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case ICE:
            case GRASS:
            case BUG:
            case STEEL:
                return VERY_EFFECTIVE;
            case FIRE:
            case WATER:
            case ROCK:
            case DRAGON:
                return NOT_EFFECTIVE;
            default: return NORMAL;
        }
    }

    public static final int getWaterAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case FIRE:
            case GROUND:
            case ROCK:
                return VERY_EFFECTIVE;
            case WATER:
            case GRASS:
            case DRAGON:
                return NOT_EFFECTIVE;
            default: return NORMAL;
        }
    }

    public static final int getElectricAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case WATER:
            case FLYING:
                return VERY_EFFECTIVE;
            case ELECTRIC:
            case GRASS:
            case DRAGON:
                return NOT_EFFECTIVE;
            case GROUND:
                return NULL;
            default: return NORMAL;
        }
    }

    public static final int getGrassAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case WATER:
            case GROUND:
            case ROCK:
                return VERY_EFFECTIVE;
            case FIRE:
            case GRASS:
            case POISON:
            case FLYING:
            case BUG:
            case DRAGON:
            case STEEL:
                return NOT_EFFECTIVE;
            default: return NORMAL;
        }
    }

    public static final int getIceAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case GRASS:
            case GROUND:
            case FLYING:
            case DRAGON:
                return VERY_EFFECTIVE;
            case FIRE:
            case WATER:
            case ICE:
            case STEEL:
                return NOT_EFFECTIVE;
            default: return NORMAL;
        }
    }

    public static final int getFightingAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case NORMAL:
            case ICE:
            case ROCK:
            case DARK:
            case STEEL:
                return VERY_EFFECTIVE;
            case POISON:
            case FLYING:
            case PSYCHIC:
            case BUG:
            case FAIRY:
                return NOT_EFFECTIVE;
            case GHOST:
                return NULL;
            default: return NORMAL;
        }
    }

    public static final int getPoisonAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case GRASS:
            case FAIRY:
                return VERY_EFFECTIVE;
            case POISON:
            case GROUND:
            case ROCK:
            case GHOST:
                return NOT_EFFECTIVE;
            case STEEL:
                return NULL;
            default: return NORMAL;
        }
    }

    public static final int getGroundAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case FIRE:
            case ELECTRIC:
            case POISON:
            case ROCK:
            case STEEL:
                return VERY_EFFECTIVE;
            case GRASS:
            case BUG:
                return NOT_EFFECTIVE;
            case FLYING:
                return NULL;
            default: return NORMAL;
        }
    }

    public static final int getFlyingAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case GRASS:
            case FIGHTING:
            case BUG:
                return VERY_EFFECTIVE;
            case ELECTRIC:
            case ROCK:
            case STEEL:
                return NOT_EFFECTIVE;
            default: return NORMAL;
        }
    }

    public static final int getPsychicAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case FIGHTING:
            case POISON:
                return VERY_EFFECTIVE;
            case PSYCHIC:
            case STEEL:
                return NOT_EFFECTIVE;
            case DARK:
                return NULL;
            default: return NORMAL;
        }
    }

    public static final int getBugAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case GRASS:
            case PSYCHIC:
            case DARK:
                return VERY_EFFECTIVE;
            case FIGHTING:
            case FIRE:
            case POISON:
            case FLYING:
            case GHOST:
            case STEEL:
            case FAIRY:
                return NOT_EFFECTIVE;
            default: return NORMAL;
        }
    }

    public static final int getRockAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case FIRE:
            case ICE:
            case BUG:
            case FLYING:
                return VERY_EFFECTIVE;
            case FIGHTING:
            case GROUND:
            case STEEL:
                return NOT_EFFECTIVE;
            default: return NORMAL;
        }
    }

    public static final int getGhostAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case PSYCHIC:
            case GHOST:
                return VERY_EFFECTIVE;
            case DARK:
                return NOT_EFFECTIVE;
            case NORMAL:
                return NULL;
            default: return NORMAL;
        }
    }

    public static final int getDragonAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case DRAGON:
                return VERY_EFFECTIVE;
            case STEEL:
                return NOT_EFFECTIVE;
            case FAIRY:
                return NULL;
            default: return NORMAL;
        }
    }

    public static final int getDarkAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case PSYCHIC:
            case GHOST:
                return VERY_EFFECTIVE;
            case FIGHTING:
            case DARK:
            case FAIRY:
                return NOT_EFFECTIVE;
            default: return NORMAL;
        }
    }

    public static final int getSteelAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case ICE:
            case ROCK:
            case FAIRY:
                return VERY_EFFECTIVE;
            case FIRE:
            case WATER:
            case ELECTRIC:
            case STEEL:
                return NOT_EFFECTIVE;
            default: return NORMAL;
        }
    }

    public static final int getFairyAttackResult(PokemonType defenderType) {
        switch (defenderType) {
            case FIGHTING:
            case DRAGON:
            case DARK:
                return VERY_EFFECTIVE;
            case FIRE:
            case POISON:
            case STEEL:
                return NOT_EFFECTIVE;
            default: return NORMAL;
        }
    }
}

File: core/src/main/java/org/zenika/core/PokemonUtils.kt
package org.zenika.core

import java.util.*


const val NORMAL = 2
const val NULL = 0
const val NOT_EFFECTIVE = 1
const val VERY_EFFECTIVE = 4

object PokemonUtils {

    private val random = Random()

    private fun randomPickType(pokemon: Pokemon): PokemonType {
        val types = pokemon.types
        val pickedType = types.pick()
        return PokemonType.fromValue(pickedType)
    }

    @JvmStatic
    fun <T> List<T>.pick(): T = this[random.nextInt(size)]

    @JvmOverloads
    @JvmStatic
    fun getAttackResult(
            attacker: Pokemon,
            defender: Pokemon,
            pickType: (Pokemon) -> PokemonType = PokemonUtils::randomPickType
    ): Int {
        val attackerType = pickType(attacker)
        val defenderType = pickType(defender)
        return when (attackerType) {
            PokemonType.POISON   -> getPoisonAttackResult(defenderType)
            PokemonType.GRASS    -> getGrassAttackResult(defenderType)
            PokemonType.FIRE     -> getFireAttackResult(defenderType)
            PokemonType.FLYING   -> getFlyingAttackResult(defenderType)
            PokemonType.WATER    -> getWaterAttackResult(defenderType)
            PokemonType.BUG      -> getBugAttackResult(defenderType)
            PokemonType.NORMAL   -> getNormalAttackResult(defenderType)
            PokemonType.ELECTRIC -> getElectricAttackResult(defenderType)
            PokemonType.GROUND   -> getGroundAttackResult(defenderType)
            PokemonType.FAIRY    -> getFairyAttackResult(defenderType)
            PokemonType.FIGHTING -> getFightingAttackResult(defenderType)
            PokemonType.PSYCHIC  -> getPsychicAttackResult(defenderType)
            PokemonType.ROCK     -> getRockAttackResult(defenderType)
            PokemonType.STEEL    -> getSteelAttackResult(defenderType)
            PokemonType.ICE      -> getIceAttackResult(defenderType)
            PokemonType.GHOST    -> getGhostAttackResult(defenderType)
            PokemonType.DRAGON   -> getDragonAttackResult(defenderType)
            PokemonType.DARK     -> getDarkAttackResult(defenderType)
        }
    }

    private fun getNormalAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.ROCK, PokemonType.STEEL -> NOT_EFFECTIVE
            PokemonType.GHOST                   -> NULL
            else                                -> NORMAL
        }
    }

    private fun getFireAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.ICE, PokemonType.GRASS, PokemonType.BUG, PokemonType.STEEL    -> VERY_EFFECTIVE
            PokemonType.FIRE, PokemonType.WATER, PokemonType.ROCK, PokemonType.DRAGON -> NOT_EFFECTIVE
            else                                                                      -> NORMAL
        }
    }

    private fun getWaterAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.FIRE, PokemonType.GROUND, PokemonType.ROCK   -> VERY_EFFECTIVE
            PokemonType.WATER, PokemonType.GRASS, PokemonType.DRAGON -> NOT_EFFECTIVE
            else                                                     -> NORMAL
        }
    }

    private fun getElectricAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.WATER, PokemonType.FLYING                       -> VERY_EFFECTIVE
            PokemonType.ELECTRIC, PokemonType.GRASS, PokemonType.DRAGON -> NOT_EFFECTIVE
            PokemonType.GROUND                                          -> NULL
            else                                                        -> NORMAL
        }
    }

    private fun getGrassAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.WATER, PokemonType.GROUND, PokemonType.ROCK                                                                             -> VERY_EFFECTIVE
            PokemonType.FIRE, PokemonType.GRASS, PokemonType.POISON, PokemonType.FLYING, PokemonType.BUG, PokemonType.DRAGON, PokemonType.STEEL -> NOT_EFFECTIVE
            else                                                                                                                                -> NORMAL
        }
    }

    private fun getIceAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.GRASS, PokemonType.GROUND, PokemonType.FLYING, PokemonType.DRAGON -> VERY_EFFECTIVE
            PokemonType.FIRE, PokemonType.WATER, PokemonType.ICE, PokemonType.STEEL       -> NOT_EFFECTIVE
            else                                                                          -> NORMAL
        }
    }

    private fun getFightingAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.NORMAL, PokemonType.ICE, PokemonType.ROCK, PokemonType.DARK, PokemonType.STEEL      -> VERY_EFFECTIVE
            PokemonType.POISON, PokemonType.FLYING, PokemonType.PSYCHIC, PokemonType.BUG, PokemonType.FAIRY -> NOT_EFFECTIVE
            PokemonType.GHOST                                                                               -> NULL
            else                                                                                            -> NORMAL
        }
    }

    private fun getPoisonAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.GRASS, PokemonType.FAIRY                                        -> VERY_EFFECTIVE
            PokemonType.POISON, PokemonType.GROUND, PokemonType.ROCK, PokemonType.GHOST -> NOT_EFFECTIVE
            PokemonType.STEEL                                                           -> NULL
            else                                                                        -> NORMAL
        }
    }

    private fun getGroundAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.FIRE, PokemonType.ELECTRIC, PokemonType.POISON, PokemonType.ROCK, PokemonType.STEEL -> VERY_EFFECTIVE
            PokemonType.GRASS, PokemonType.BUG                                                              -> NOT_EFFECTIVE
            PokemonType.FLYING                                                                              -> NULL
            else                                                                                            -> NORMAL
        }
    }

    private fun getFlyingAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.GRASS, PokemonType.FIGHTING, PokemonType.BUG  -> VERY_EFFECTIVE
            PokemonType.ELECTRIC, PokemonType.ROCK, PokemonType.STEEL -> NOT_EFFECTIVE
            else                                                      -> NORMAL
        }
    }

    private fun getPsychicAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.FIGHTING, PokemonType.POISON -> VERY_EFFECTIVE
            PokemonType.PSYCHIC, PokemonType.STEEL   -> NOT_EFFECTIVE
            PokemonType.DARK                         -> NULL
            else                                     -> NORMAL
        }
    }

    private fun getBugAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.GRASS, PokemonType.PSYCHIC, PokemonType.DARK                                                                                -> VERY_EFFECTIVE
            PokemonType.FIGHTING, PokemonType.FIRE, PokemonType.POISON, PokemonType.FLYING, PokemonType.GHOST, PokemonType.STEEL, PokemonType.FAIRY -> NOT_EFFECTIVE
            else                                                                                                                                    -> NORMAL
        }
    }

    private fun getRockAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.FIRE, PokemonType.ICE, PokemonType.BUG, PokemonType.FLYING -> VERY_EFFECTIVE
            PokemonType.FIGHTING, PokemonType.GROUND, PokemonType.STEEL            -> NOT_EFFECTIVE
            else                                                                   -> NORMAL
        }
    }

    private fun getGhostAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.PSYCHIC, PokemonType.GHOST -> VERY_EFFECTIVE
            PokemonType.DARK                       -> NOT_EFFECTIVE
            PokemonType.NORMAL                     -> NULL
            else                                   -> NORMAL
        }
    }

    private fun getDragonAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.DRAGON -> VERY_EFFECTIVE
            PokemonType.STEEL  -> NOT_EFFECTIVE
            PokemonType.FAIRY  -> NULL
            else               -> NORMAL
        }
    }

    private fun getDarkAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.PSYCHIC, PokemonType.GHOST                    -> VERY_EFFECTIVE
            PokemonType.FIGHTING, PokemonType.DARK, PokemonType.FAIRY -> NOT_EFFECTIVE
            else                                                      -> NORMAL
        }
    }

    private fun getSteelAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.ICE, PokemonType.ROCK, PokemonType.FAIRY                         -> VERY_EFFECTIVE
            PokemonType.FIRE, PokemonType.WATER, PokemonType.ELECTRIC, PokemonType.STEEL -> NOT_EFFECTIVE
            else                                                                         -> NORMAL
        }
    }

    private fun getFairyAttackResult(defenderType: PokemonType): Int {
        return when (defenderType) {
            PokemonType.FIGHTING, PokemonType.DRAGON, PokemonType.DARK -> VERY_EFFECTIVE
            PokemonType.FIRE, PokemonType.POISON, PokemonType.STEEL    -> NOT_EFFECTIVE
            else                                                       -> NORMAL
        }
    }
}
